trio/_abc.py

```diff
@@ -483,7 +484,22 @@ class HalfCloseableStream(Stream):
         """
 
 
-class Listener(AsyncResource):
+# The type of object produced by a ReceiveChannel (covariant because
+# ReceiveChannel[Derived] can be passed to someone expecting
+# ReceiveChannel[Base])
+T_co = TypeVar("T_co", covariant=True)
+
+# The type of object accepted by a SendChannel (contravariant because
+# SendChannel[Base] can be passed to someone expecting
+# SendChannel[Derived])
+T_contra = TypeVar("T_contra", contravariant=True)
+
+# The type of object produced by a Listener (covariant plus must be
+# an AsyncResource)
+T_resource = TypeVar("T_resource", bound=AsyncResource, covariant=True)
```

Comment:
I think the `bound=AsyncResource` will disappear when we fix #636. We could start skipping it now, or we could put it in and then take it out later, doesn't make much difference IMO.

Reply:
You think we'll have Listeners that provide objects not representable as an AsyncResource? I'm a bit surprised at that, can you give an example?

Reply:
I don't have an example in mind. But the reason we restrict it to `AsyncResource` right now is because `serve_listeners` automatically closes the object after the handler returns: https://github.com/python-trio/trio/blob/0a815a015bd06fa6ff929d3bb881c9d44e6d9c8a/trio/_highlevel_serve_listeners.py#L25-L29

With #636, this code goes away, because the `Listener` itself is responsible for managing the handler lifetime. So... at that point it's really between the person developing a `Listener[X]` and their users, what kind of API `X` should support :-).

Reply:
Ah, that makes sense! I think it's probably sensible to keep the AsyncResource bound for now, since it reflects the current requirements, and we can easily change it later.